### PR TITLE
Add aws plan command to AWS Provider

### LINKED ISSUE #7449: Add plan command to AWS provider
# Feature Proposal
Add serverless plan command to AWS provider

## Description
It would be useful to have a plan command that shows a detailed log of the changes from a CloudFormation changeset. Similar to `terraform plan`, see https://www.terraform.io/docs/commands/plan.html

#  What is the use case that should be solved?
As a developer, I want to see a detailed log of changes before executing the deployment to AWS CloudFormation so that I don't deploy unintended changes.

# Implementation proposal
- create a new AWS provider-specific command called plan
- when plan is called then do the following
  - spawn serverless deploy preventing the execution of the stack update
  - create a changeset from the stack update
  - print the changeset details to the console
  - destroy the changeset and its corresponding files

# Contribution
There's a fork that I'd like to use to gather feedback before creating a pull request as recommended in the contribution guidelines. Changes can be reviewed in the last commit of the branch at https://github.com/edgarpoce/serverless/commits/feature/awsplan

Code from other open source projects was included and refactored as necessary. These projects are distributed under MIT license and the refactored files contain the corresponding copyright notice
https://github.com/trek10inc/serverless-cloudformation-changesets 
https://github.com/sjakthol/cfn-execute-change-set

# Similar issues:

- #4179

### PR DESCRIPTION
**Pull request by edgarpoce**
## What did you implement
Add serverless plan command to AWS provider

- create a new AWS provider-specific command called plan
- lifecycle
  - 'package', spawn sls package
  - 'checkForChanges', from deploy command (cancel if no changes)
  - 'uploadCloudFormationFile', from deploy command
   - 'createChangeSet', new module
  - 'waitForChangeSetCreateComplete', new module
  - 'runAnalysis', new module
  - 'deleteChangeSet', new module
  - 'deleteChangeSetFromS3', new module
  - 'printAnalysis', new module

Closes #7449

## How can we verify it

Run 
`sls deploy` on any sls project.
`sls plan` afterwards

## serverless.yml sample
```
provider:
  name: aws
  runtime: nodejs10.x

functions:
  helloWorld:
    handler: src/helloWorld.handler
    events:
      - http:
          path: hello-world
          method: get
          cors: true

```

## Todos

<details>
<summary>Useful Scripts</summary>

- `npm run test:ci` --> Run all validation checks on proposed changes
- `npm run lint:updated` --> Lint all the updated files
- `npm run lint:fix` --> Automatically fix lint problems (if possible)
- `npm run prettier-check:updated` --> Check if updated files adhere to Prettier config
- `npm run prettify:updated` --> Prettify all the updated files
</details>

- [x] Write and run all tests
- [x] Write documentation
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

### CODE REVIEW

File: lib/plugins/aws/plan/lib/naming.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+function getChangeSetS3FolderPath(plugin) {
+  return plugin.serverless.service.package.artifactDirectoryName;
+}
+
+function getChangeSetS3FolderUrl(plugin) {
+  const bucketName = getS3BucketName(plugin);
+  const s3Folder = getChangeSetS3FolderPath(plugin);
+  return `https://s3.amazonaws.com/${bucketName}/${s3Folder}`;
```
Comment (maafk) [inline lib/plugins/aws/plan/lib/naming.js]: For certain regions, `s3.amazonaws.com` is invalid.

Could use the [`getS3EndpointForRegion()`](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/utils/getS3EndpointForRegion.js) function to replace this.

```javascript
'use strict';
const getS3EndpointForRegion = require('../../utils/getS3EndpointForRegion');


function getChangeSetS3FolderUrl(plugin) {
  const bucketName = getS3BucketName(plugin);
  const s3Folder = getChangeSetS3FolderPath(plugin);
  const s3Endpoint = getS3EndpointForRegion(plugin.provider.getRegion());
  return `https://${s3Endpoint}/${bucketName}/${s3Folder}`;
}

```

Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/naming.js:22]: nice catch!, will fix it ASAP. Thanks!

File: lib/plugins/aws/plan/index.test.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    it('should set the serverless instance', () => expect(awsPlan.serverless).to.equal(serverless));
+
+    it('should set options', () => expect(awsPlan.options).to.equal(options));
+
+    it('should set the provider variable to an instance of AwsProvider', () =>
+      expect(awsPlan.provider).to.be.instanceof(AwsProvider));
+
+    it('should have commands', () => expect(awsPlan.commands).to.be.not.empty);
+
+    it('should have hooks', () => expect(awsPlan.hooks).to.be.not.empty);
```
Comment (medikoo) [inline lib/plugins/aws/plan/index.test.js:37]: Currently we're trying to not introduce such tests for new functionality (sorry that it's not properly communicated at this point).

Problem with tests like that, is that they do not really test outcome (so what's important) but characteristics of internal implementation, which should rather be treated as a black box, and such tests make eventual future refactor really painful.

So instead of that, we now use [`runServerless`](https://github.com/serverless/test/blob/master/docs/run-serverless.md) util, which initialize complete `serverless` instance (against pre-configured service, which can be passed either via inline config or pre-prepared fixture), and test the outcome of that.

Good examples to check:
- https://github.com/serverless/serverless/blob/80f1c798a293d99529f9e7200227aa0f96ace219/lib/plugins/aws/package/compile/events/httpApi/index.test.js
- https://github.com/serverless/serverless/blob/80f1c798a293d99529f9e7200227aa0f96ace219/lib/plugins/aws/package/compile/functions/index.test.js#L2624-L2718

As when testing `plan` you'd need to mock AWS requests, see how it's done here: https://github.com/serverless/serverless/blob/656c0c5abc3748b10efbacc753ea536bc8ca773a/lib/plugins/aws/deploy/lib/uploadArtifacts.test.js#L18-L39

Ideally for plan, we should have one test file `plan/index.test.js` (no need to create test files for all lib files), in which we should test most common _plan_ scenarios (no need to not dive too much into corner cases), by confirming on generated output which as I can can be read from `plugin.analysis`.

If you have any doubts, questions, please ask. I don't want this to feel as blocker

File: lib/plugins/aws/plan/lib/createChangeSet.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
+// copyright Copyright (c) 2017 Trek10
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/createChangeSet.js:1]: I added a note in the issue commenting that I included code from other open source projects. is this OK or is there any other guideline for these cases?

> Code from other open source projects was included and refactored as necessary. These projects are distributed under MIT license and the refactored files contain the corresponding copyright notice
https://github.com/trek10inc/serverless-cloudformation-changesets
https://github.com/sjakthol/cfn-execute-change-set

**Comment by edgarpoce** [comment]
@medikoo , many thanks for reviewing it!. I'll tackle the change requests ASAP and will update the branch. thanks!

File: lib/plugins/aws/plan/lib/createChangeSet.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (medikoo) [reply lib/plugins/aws/plan/lib/createChangeSet.js:1]: Why can't we just require them as dependencies in such case?

[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/createChangeSet.js:1]: that's was my initial approach but unfortunately none of these two projects seem designed to be reused. I refactored pieces of both to make the logic usable in the context of this contribution.

[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/createChangeSet.js:1]: I'll remove the notes to avoid confusion, I assume it's OK because those two projects are licensed under MIT license

[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/createChangeSet.js]: comments removed in https://github.com/serverless/serverless/pull/7509/commits/cca1c0bf9d92c73276b497aa5bb8d6e1a89f18bc

File: lib/plugins/aws/plan/lib/naming.js (hunk 1 above)
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/naming.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/7eac973c4c71f17911c985178c2a5c94f3419fba

File: lib/plugins/aws/plan/index.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+          'deleteChangeSet',
+          'deleteChangeSetFromS3',
+          'printAnalysis',
+        ],
+      },
+    };
+
+    this.hooks = {
+      'before:plan:package': () =>
+        BbPromise.bind(this)
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/index.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/8945eb722d3dd4488c58e7751c927bf6845cf4c2

File: lib/plugins/aws/plan/lib/createChangeSet.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+function createChangeSet(changeSetSuffix) {
+  changeSetSuffix = changeSetSuffix || Date.now();
+  const stackName = this.provider.naming.getStackName();
+  const changeSetName = `${stackName}-${changeSetSuffix}`;
+  this.changeSetName = changeSetName;
+  this.serverless.cli.log(`Creating ChangeSet [${changeSetName}]`);
+  return execute(this, 'UPDATE').catch(e => {
+    if (e.message.indexOf('does not exist') > -1) {
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/createChangeSet.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/0b0c165ec09b195d2737f86ac01730a49eabeee3

File: lib/plugins/aws/plan/lib/analysisPrinter.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2017 Sami Jaktholm <sjakthol@outlook.com>
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/analysisPrinter.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/cca1c0bf9d92c73276b497aa5bb8d6e1a89f18bc

File: lib/plugins/aws/plan/lib/analysisRenderers.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2017 Sami Jaktholm <sjakthol@outlook.com>
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/analysisRenderers.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/cca1c0bf9d92c73276b497aa5bb8d6e1a89f18bc

File: lib/plugins/aws/plan/lib/deleteChangeSet.test.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2017 Sami Jaktholm <sjakthol@outlook.com>
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/deleteChangeSet.test.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/cca1c0bf9d92c73276b497aa5bb8d6e1a89f18bc

File: lib/plugins/aws/plan/lib/printAnalysis.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+'use strict';
+function printAnalysis() {
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/printAnalysis.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/ecab9f22e9d81362d6fba7c0c2513c7d405e6b31

File: lib/plugins/aws/plan/lib/naming.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
+'use strict';
+
+function getStackName(plugin) {
+  return plugin.provider.naming.getStackName();
+}
+
+function getChangeSetName(plugin) {
+  return plugin.changeSetName;
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/naming.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/94536d17042c2a99d79091049384066983736445

File: lib/plugins/aws/plan/index.js [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+      'plan:package': () => {
+        return this.serverless.pluginManager.spawn('package');
+      },
+      'plan:checkForChanges': () =>
+        BbPromise.bind(this)
+          .then(this.checkForChanges)
+          .then(() => {
+            if (this.serverless.service.provider.shouldNotDeploy) {
+              return BbPromise.reject(
+                new this.serverless.classes.Error('Service files not changed')
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/index.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/15435bdfc273fd5119fce8f6ff4a88fb9124b338

File: lib/plugins/aws/plan/lib/createChangeSet.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+  const stackName = naming.getStackName(plugin);
+  const changeSetName = naming.getChangeSetName(plugin);
+  const templateUrl = naming.getChangeSetS3CompiledTemplateUrl(plugin);
+
+  let stackTags = {
+    STAGE: plugin.provider.getStage(),
+  };
+
+  // Merge additional stack tags
+  if (typeof plugin.serverless.service.provider.stackTags === 'object') {
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/createChangeSet.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/f5ed2d9fa71a8e88c67e8683988dab2d881ff88f

File: lib/plugins/aws/plan/lib/createChangeSet.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    stackTags = _.extend(stackTags, plugin.serverless.service.provider.stackTags);
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/createChangeSet.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/696042b66065d206cf1c70f89f8aad523944ad5e

File: lib/plugins/aws/plan/lib/runAnalysis.js [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+  const region = this.provider.getRegion();
+  const stackName = naming.getStackName(this);
+  const changeSetName = naming.getChangeSetName(this);
+  this.analysis = '';
+  const logger = {
+    log: msg => {
+      this.analysis += `${msg}\n`;
+    },
+  };
+  return BbPromise.join(
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/runAnalysis.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/5994f266027db6e235297a6534923def6d01e9f5

File: lib/plugins/aws/plan/lib/analysis.js [hunk 15]
```diff
@@ -0,0 +1,271 @@
+// Copyright (c) 2017 Sami Jaktholm <sjakthol@outlook.com>
+'use strict';
+const _ = require('lodash');
+
+const lo = {
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/analysis.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/dedd322c9184aa2edbb363845b69b5f9e190b34e

File: lib/plugins/aws/plan/lib/waitForChangeSetCreateComplete.js [hunk 16]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+
+const AWS = require('aws-sdk');
+const naming = require('./naming');
+
+function waitForChangeSetCreateComplete() {
+  const credentials = Object.assign({}, this.provider.getCredentials());
+  credentials.region = this.provider.getRegion();
+  const cf = new AWS.CloudFormation(credentials);
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/waitForChangeSetCreateComplete.js:9]: Hi @medikoo, 

I tried to use provider.request but found that the implementation doesn't support passing more than one parameter, it accepts only a single object that is passed to the aws request. 

Unfortunately cloudformation.waitFor doesn't follow the same structure than most AWS methods. Cloudformation.waitFor takes two parameters, see https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudFormation.html#waitFor-property.

I think we can either keep the implementation as it is or modify provider.request. What's your preference to tackle this issue?

File: lib/plugins/aws/plan/lib/runAnalysis.js [hunk 17]
```diff
@@ -0,0 +1,37 @@
+'use strict';
+const BbPromise = require('bluebird');
+const printer = require('./analysisPrinter');
+const naming = require('./naming');
+
+function describeStack(provider, stackName, region) {
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/runAnalysis.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/8928fc25de7ad425f952bdd69661d7cb156d1636

File: lib/plugins/aws/plan/index.js [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+              );
+            }
+            return BbPromise.resolve();
+          }),
+      'plan:uploadCloudFormationFile': () =>
+        BbPromise.bind(this).then(this.uploadCloudFormationFile),
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/index.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/a3634ff54ded2b75d6aa8dee07623eace46eec39

File: lib/plugins/aws/plan/lib/waitForChangeSetCreateComplete.js [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+  return cf
+    .waitFor('changeSetCreateComplete', {
+      StackName: naming.getStackName(this),
+      ChangeSetName: naming.getChangeSetName(this),
+    })
+    .promise();
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/lib/waitForChangeSetCreateComplete.js:15]: The default timeout seems quite high. Please reopen the discussion if we should modify this.

`Waits for the changeSetCreateComplete state by periodically calling the underlying CloudFormation.describeChangeSet() operation every 30 seconds (at most 120 times).`

File: lib/plugins/aws/plan/index.js [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+            this.serverless.cli.log('unable to get bucket name');
+            this.serverless.cli.log('"plan" runs only on deployed projects');
+            throw error;
+          }),
+      'plan:package': () => {
+        return this.serverless.pluginManager.spawn('package');
+      },
+      'plan:checkForChanges': () =>
+        BbPromise.bind(this)
+          .then(this.checkForChanges)
```
[the remark this answers is not in the document]
Comment (edgarpoce) [reply lib/plugins/aws/plan/index.js]: fixed in https://github.com/serverless/serverless/pull/7509/commits/f00e4e2ce5374b38683e9e8090cbeadc2bfa0caa

File: lib/plugins/aws/plan/index.test.js (hunk 2 above)
Comment (edgarpoce) [reply lib/plugins/aws/plan/index.test.js]: Many thanks for the recommendation!, it's a very useful approach!. I refactored the test of the entry point to use a slightly different version of this approach. I don't see the mechanism to mock aws, as a result I implemented a very similar mechanism that allows to use sinon to mock aws.
If you are OK I'll keep the unit tests and I'll refactor them to this kind of functional test as changes are needed or any bug is identified. 
Please let me know if this is OK or if I should remove the unit tests.
proposed fix is committed in https://github.com/serverless/serverless/pull/7509/commits/6e05e24be27735e43af1aab19fea39798e0ad51a

**Comment by edgarpoce** [comment]
@medikoo, 
I've provided fixes for the issues we discussed. I haven't renamed the plan command to facilitate the review, just to keep files in the same folder and avoid creating a big diff. I'll rename the command as soon as other changes are reviewed. 
Many thanks for reviewing!

**Comment by troyready** [comment]
I can appreciate the value & need of breaking up the monolith.

What would a plugin look like in this context -- like a traditional serverless plugin? Or something automatically included in the serverless AWS provider?

Really, what I'm wondering is if the new command would be available ootb for AWS provider users? Looking at the docs now, I can see a command like [metrics](https://serverless.com/framework/docs/providers/aws/cli-reference/metrics/) that's only available on AWS, and commands like `deploy` that are on [AWS](https://serverless.com/framework/docs/providers/aws/cli-reference/deploy/) and [Azure.](https://serverless.com/framework/docs/providers/azure/cli-reference/deploy/)

Are we talking about an implementation detail for getting the command in for AWS users by default (and additional providers as we add support -- I can speak for Edgar in saying that he'd likely have time to work on the Azure one too if the APIs support it), so there'd be a new `plan` command under the AWS CLI provider docs?

Or would it be something that users would have to install separately?

**Comment by medikoo** [comment]
> What would a plugin look like in this context -- like a traditional serverless plugin? Or something automatically included in the serverless AWS provider?

It won't be automatically included (otherwise it'll have to come with the core). It'll work as other plugins listed here: https://github.com/serverless/plugins

So will have to be installed seperately and listed in `serverless.yml` plugins section.

> and commands like deploy that are on AWS and Azure.

It's only AWS that's baked in. Azure users needs to install [serverless-azure-functions](https://github.com/serverless/serverless-azure-functions) plugin.
We have though Azure (and other providers) documentation included in this repository, but we regret that. It should rather be kept at provider plugins (and it's likely we switch to that in near future)

We also look into direction where AWS support will be taken out from main repo and split into distinct plugins. Users then will deal with light setup and just take what they need.

**Comment by troyready** [comment]
I can appreciate the technical merits of moving to a plugin-based approached to the different services. It sounds akin to what Terraform does with their provider system, which in my experience has brought more benefits than drawbacks.

For a core command like this, I don't think it makes sense to distribute it separately from the other AWS support. Unless a technical lesson has been learned with Serverless Framework that k8s (`diff`), AWS CDK (`diff`), Terrafom (`plan`), etc have missed, it seems more sensible to bring sls up to parity with other systems.

**Comment by medikoo** [comment]
@edgarpoce @troyready We've discussed this internally, and decided to not include it in Framework core, but provide it as external plugin.

I totally agree that it can be seen as a _core_ feature for AWS CloudFront based deployment, in sense that it technically can be used with any kind of AWS deployment as done with a Framework.
However it doesn't really integrate with _core_ functionality, and brings no aid to it. Technically it's something added aside, which is notable in size and for which demand seems low, hence can totally be kept as external plugin.

As I mentioned, we're now very savvy in adding new things, as Framework is already bloated and difficult to maintain. 

Anyway you've done a great work here, and I'm sure many developers would love to benefit from that. We'd like to provide a room for it under [serverless](https://github.com/serverless/) organization at e.g. `serverless-plugin-cloudformation-diff` repository, and help with maintenance of it.  Let us know if that works for you (?)

**Comment by edgarpoce** [comment]
Hi @medikoo , many thanks for taking the time to review the work and share your insights!. Our intention with this contribution was to be able to invoke cloudformation-diff from an external tool without the requirement of declaring the dependency in the serverless project. unless I'm missing something moving the contribution out of core would prevent this goal.

is there any functionality or workaround to include external plugins from the command line?, 
e.g  `sls cloudformation-diff --plugins @serverless/cloudformation-diff. `
If not available, is this kind of functionality aligned with the serverless framework?, Not sure, but I think we might be able to allocate some time to implement it if there's interest. WDYT?

**Comment by medikoo** [comment]
@edgarpoce we have dedicated internal command for adding plugins. So technically once it's ported as external plugin. the following should work:

```
$ serverless plugin install cloudformation-diff
...
$ sls cloudformation-dff
...
```
